### PR TITLE
Stronger checks for invalid sync codes

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Resolves https://github.com/brave/brave-browser/issues/19200

## Submitter Checklist:

- [?] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Test plan is intended to ensure that we will not got the CHECK (crashes) while regular sync operations.

I. Android

1. Create a new chain, ensure device is in chain
2. Exit browser
3. Start browser, ensure device is in chain
4. Leave sync chain
5. Enter the sync by QR code, ensure device is in chain
6. Leave sync chain
7. Enter the sync by passphrase, ensure device is in chain

II. Desktop - quite the same as Android

1. Create a new chain, ensure device is in chain
2. Exit browser
3. Start browser, ensure device is in chain
4. Leave sync chain
5. Enter the sync by passphrase, ensure device is in chain

### CODE REVIEW

File: components/brave_sync/crypto/crypto.cc [hunk 1]
```diff
@@ -151,7 +151,7 @@ bool PassphraseToBytes32(const std::string& passphrase,
   bytes->resize(DEFAULT_SEED_SIZE);
   if (bip39_mnemonic_to_bytes(nullptr, passphrase.c_str(), bytes->data(),
                               bytes->size(), &written) != WALLY_OK) {
-    LOG(ERROR) << "bip39_mnemonic_to_bytes failed";
+    CHECK(false) << "bip39_mnemonic_to_bytes failed";
```
Comment (darkdh) [inline components/brave_sync/crypto/crypto.cc:154]: I think you meant to put this in `PassphraseFromBytes32`
The change you did here will crash the browser when UI pass a invalid mnemonic, and UI relies on it to tell users the sync words are invalid.

Comment (AlexeyBarabash) [reply components/brave_sync/crypto/crypto.cc:154]: @darkdh , yes, thanks, I will revert this because you are 100% right

Comment (AlexeyBarabash) [reply components/brave_sync/crypto/crypto.cc]: force-pushed at https://github.com/brave/brave-core/compare/c91ceb642e45dcfbbdf82e945b54eb5032dc10a1..b8715733f07249e053329251c7681db46182af1b

File: components/brave_sync/crypto/crypto_unittest.cc [hunk 2]
```diff
@@ -239,6 +241,9 @@ TEST(CryptoTest, Passphrase) {
       "mountain jelly";
   EXPECT_TRUE(PassphraseToBytes32(bip_passphrase, &bip_bytes));
   EXPECT_EQ(PassphraseFromBytes32(bip_bytes), bip_passphrase);
+  EXPECT_TRUE(IsPassphraseValid(bip_passphrase));
+  EXPECT_FALSE(IsPassphraseValid(""));
+  EXPECT_FALSE(IsPassphraseValid(bip_passphrase + " something wrong"));
```
Comment (fmarier) [inline components/brave_sync/crypto/crypto_unittest.cc:246]: One more case I would suggest is:
```
const std::string bip_invalid_passphrase =
      "magic vacuum wide review love peace century egg burden clutch heart "
      "cycle annual mixed pink awesome extra client cry brisk priority maple "
      "mountain brave";
EXPECT_FALSE(IsPassphraseValid(bip_invalid_passphrase));
```
(I kept the length the same, but replaced the last word with an invalid word -- `brave`.)

**Review by fmarier — APPROVED** [review]

File: components/brave_sync/crypto/crypto_unittest.cc (hunk 2 above)
Comment (AlexeyBarabash) [reply components/brave_sync/crypto/crypto_unittest.cc:250]: thanks @fmarier , updated in https://github.com/brave/brave-core/pull/10960/commits/69caa7fe0b3566c3f743291a7ac2509ae17c87c3#diff-b912644143e341afc9dc952720124c3876ff375bb276ca70f9c90346c474759fR242

**Review by darkdh — APPROVED** [review]

**Comment by AlexeyBarabash** [comment]
Windows CI is marked `UNSTABLE`, not related to the PR:
https://ci.brave.com/job/pr-brave-browser-sync_codes_stronger_check-windows/3/execution/node/316/log/
```
00:49:29  Uploading file:/C:/heck/src/out/Release/brave_installer_96_1_34_2.exe to s3://****/brave-browser-build-pr/sync_codes_stronger_check/207c4d79be1c64ee72901bc82b874d1998cdae34-7ab466de207833426b1f8b90f88ec576ff56c71f/windows/ 
00:49:29  Upload failed due to missing source file
```

**Comment by AlexeyBarabash** [comment]
`Post-init` CI step is still of `is being built` on GitHub, but it is completely green in fact:
https://ci.brave.com/job/pr-brave-browser-sync_codes_stronger_check-noplatform/5//flowGraphTable/?auto_refresh=true

**Comment by AlexeyBarabash** [comment]
Verified on own-built `master`

**Comment by stephendonner** [comment]
Verified `PASSED` using the testplan here on build 

Brave | 1.34.9 Chromium: 96.0.4664.35 (Official Build) nightly (x86_64)
-- | --
Revision | 8df3babf8da8faa8aacb78f4848ef66a93b4944b-refs/branch-heads/4664@{#680}
OS | macOS Version 11.6.1 (Build 20G224)

Steps:

1. new profile
2. launched Brave
3. loaded `brave://settings/braveSync/setup` and chose `Start a new Sync Chain`
4. copied code words into clipboard
5. shut down Brave
6. loaded `brave://settings/braveSync/setup` and confirmed my device was still on the Sync chain
7. clicked on `Leave Sync Chain`
8. clicked `OK`
9. loaded `brave://settings/braveSync/setup` and clicked on `I have a Sync code`
10. entered the code words from step 4 and clicked `Confirm`
11. confirmed via `brave://settings/braveSync/setup` my device is now again on the original chain created in step 3

example | example | example | example | example | example
---------|----------|----------|---------|---------|----------
Screen Shot 2021-11-11 at 11 47 36 PM | Screen Shot 2021-11-11 at 11 48 22 PM | Screen Shot 2021-11-11 at 11 48 37 PM | Screen Shot 2021-11-11 at 11 50 27 PM | Screen Shot 2021-11-11 at 11 51 36 PM | Screen Shot 2021-11-11 at 11 51 59 PM

**Comment by stephendonner** [comment]
Verified `PASSED` using Brave 

Steps:
1. new profile
2. launch Brave
3. tapped on `hamburger` menu item on bottom toolbar
4. tapped `Settings`
5. tapped `Sync`
6. tapped `Start a new sync chain`
7. tapped `Add a Computer`
8. tapped `Copy to Clipboard` to get the Sync chain code/code words into....the clipboard
9. tapped `Done`
10. pasted code words/chain code into desktop Sync client (follow setup above)
11. confirmed desktop client joined Sync chain via `brave://settings/braveSync/setup`
12. shut down Brave
13. relaunched Brave, ensure device was still in the chain
14. tapped on the `trashcan` icon to `delete` / `remove` device from Sync chain
15. tapped `OK` when asked to confirm
16. scanned QR code from desktop and ensured device (re)joined the Sync chain
17. again tapped on the `trashcan` icon to the right of my Android device
18. tapped `OK` when asked to confirm
19. tapped on `Sync`
20. tapped on `Scan or enter sync code`
21. tapped `Enter code words`
22. pasted the code words
23. tapped `Confirm`
24. tapped `Yes` at the newly-introduced warning prompt
25. confirmed I (re)joined the Sync chain

example | example | example | example | example | example | example
---------|----------|----------|---------|----------|---------|--------
Screenshot_20211112-002315 |  Screenshot_20211112-002413 | Screenshot_20211112-003327 | Screenshot_20211112-002532 | Screenshot_20211112-002542 | Screenshot_20211112-002545 | Screenshot_20211112-003327

**Pull request merged**